### PR TITLE
`azurerm_servicebus_subscription`: support client affine property, `azurerm_servicebus_subscription_rule`: add property `sql_filter_compatibility_level`

### PR DESCRIPTION
**Pull request by xiaxyi**
support client affine property for azure servicebus subscription

client affine property for[ JMS subscriptions](https://docs.microsoft.com/en-us/azure/service-bus-messaging/java-message-service-20-entities#java-message-service-jms-subscriptions):

### CODE REVIEW

File: website/docs/d/servicebus_subscription.html.markdown [hunk 1]
```diff
@@ -57,6 +57,14 @@ output "servicebus_subscription" {
 
 * `forward_dead_lettered_messages_to` - The name of a Queue or Topic to automatically forward Dead Letter messages to.
 
+* `client_scoped_subscription_enabled` - The subscription has an affinity to the client id or not.
```
Comment (katbyte) [inline website/docs/d/servicebus_subscription.html.markdown:60]: the name here uses scoped, but the description talked about it being having an affinity? should this be
```suggestion
* `client_scoped_subscription_enabled` - The subscription has an affinity to the client id or not. 
```
```suggestion
* `client_affinity_enabled` - Does the subscription has an affinity to a client id?
```

File: website/docs/r/servicebus_subscription.html.markdown [hunk 2]
```diff
@@ -73,6 +73,18 @@ The following arguments are supported:
 
 * `status` - (Optional) The status of the Subscription. Possible values are `Active`,`ReceiveDisabled`, or `Disabled`. Defaults to `Active`.
 
+* `client_scoped_subscription_enabled` - (Optional)  whether the subscription has an affinity to the client id. Defaults to `False`. 
+
+~> **NOTE:** Client Scoped Subscription can only be used for JMS subscription.
```
Comment (katbyte) [inline website/docs/r/servicebus_subscription.html.markdown:78]: what is a JMS subscription?

File: internal/services/servicebus/servicebus_subscription_resource_test.go [hunk 3]
```diff
@@ -369,3 +384,41 @@ func (ServiceBusSubscriptionResource) updateDeadLetteringOnFilterEvaluationExcep
...
+
+resource "azurerm_servicebus_subscription" "test" {
+  name                                 = "_acctestsub-%[1]d_"
+  topic_id                             = azurerm_servicebus_topic.test.id
+  max_delivery_count                   = 10
+  auto_delete_on_idle                  = "PT5M"
+  lock_duration                        = "PT1M"
+  dead_lettering_on_message_expiration = true
+	%[3]s
+  client_scoped_subscription_enabled       = true
```
Comment (katbyte) [inline internal/services/servicebus/servicebus_subscription_resource_test.go:416]: could we clean this up? the %[3]s is confusing

File: website/docs/r/servicebus_subscription.html.markdown [hunk 4]
```diff
@@ -73,6 +73,18 @@ The following arguments are supported:
...
+
+* `client_id_for_client_scoped_subscription` - (Optional)  Specifies the Client ID of the application that created the client-scoped subscription.
+
+~> **NOTE:** Client ID can be null or empty, but it must match the client ID set on the JMS client application. From the Azure Service Bus perspective, a null client ID and an empty client id have the same behavior. If the client ID is set to null or empty, it is only accessible to client applications whose client ID is also set to null or empty.
+
+* `is_client_scoped_subscription_shareable` - (Optional) Whether the client scoped subscription is shareable. Defaults to `true`
+
+* `is_client_scoped_subscription_durable` - (Optional) Whether the client scoped subscription is durable. This property can only be controlled from the application side.
```
Comment (katbyte) [inline website/docs/r/servicebus_subscription.html.markdown:86]: would this make sense to wrap in a `client_affinity` block?
```suggestion
* `enabled` - (Optional)  whether the subscription has an affinity to a client id. Defaults to `False`. 

~> **NOTE:** Client Scoped Subscription can only be used for JMS subscription.

* `client_id` - (Optional)  Specifies the Client ID of the application that created the client-scoped subscription.

~> **NOTE:** Client ID can be null or empty, but it must match the client ID set on the JMS client application. From the Azure Service Bus perspective, a null client ID and an empty client id have the same behavior. If the client ID is set to null or empty, it is only accessible to client applications whose client ID is also set to null or empty.

* `shareable` - (Optional) Whether the client scoped subscription is shareable. Defaults to `true`

* `durable` - (Optional) Whether the client scoped subscription is durable. This property can only be controlled from the application side.
```

File: website/docs/r/servicebus_subscription.html.markdown [hunk 5]
```diff
@@ -73,6 +73,18 @@ The following arguments are supported:
 
 * `status` - (Optional) The status of the Subscription. Possible values are `Active`,`ReceiveDisabled`, or `Disabled`. Defaults to `Active`.
 
+* `client_scoped_subscription_enabled` - (Optional)  whether the subscription has an affinity to the client id. Defaults to `False`. 
+
+~> **NOTE:** Client Scoped Subscription can only be used for JMS subscription.
+
+* `client_id_for_client_scoped_subscription` - (Optional)  Specifies the Client ID of the application that created the client-scoped subscription.
```
Comment (katbyte) [inline website/docs/r/servicebus_subscription.html.markdown:80]: is this client "scoped" or client affinity?
```suggestion
* `client_id` - (Optional)  The client id this subscription should have an affinity with.
```

the mixing of terms here is quite confusing

**Review by katbyte — CHANGES_REQUESTED** [review]

File: website/docs/r/servicebus_subscription.html.markdown (hunk 2 above)
Comment (xiaxyi) [reply website/docs/r/servicebus_subscription.html.markdown:78]: Thanks Kt for the review, as the link that I put in the pr description, the full name of JMS subscription is Java message service (JMS) subscriptions: _While these are semantically similar to the [Subscriptions](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-queues-topics-subscriptions#topics-and-subscriptions) (that is, exist on a topic and enable publish/subscribe semantics), the Java Message Service spec introduces the concepts of Shared, Unshared, **Durable, and Non-durable attributes for a given subscription._

File: website/docs/d/servicebus_subscription.html.markdown (hunk 1 above)
Comment (xiaxyi) [reply website/docs/d/servicebus_subscription.html.markdown:60]: I used the word by reading the article: https://docs.microsoft.com/en-us/azure/service-bus-messaging/java-message-service-20-entities#additional-concepts-for-java-message-service-jms-20-subscriptions. The article talks about whether the subscription is scoped to a client or not.

File: internal/services/servicebus/servicebus_subscription_resource_test.go (hunk 3 above)
Comment (xiaxyi) [reply internal/services/servicebus/servicebus_subscription_resource_test.go:416]: will do

File: website/docs/r/servicebus_subscription.html.markdown (hunk 4 above)
Comment (xiaxyi) [reply website/docs/r/servicebus_subscription.html.markdown:86]: Will put the related properties into a block.

**Comment by xiaxyi** [comment]
@katbyte Thanks for the review, I will make the aligned term. Personally, I think scope is clearer, but still would like to ask for your suggestions before making any changes. Thanks!

File: website/docs/d/servicebus_subscription.html.markdown (hunk 1 above)
Comment (katbyte) [reply website/docs/d/servicebus_subscription.html.markdown:60]: thanks, i agree after reading those docs it does sound like scoped is better

**Comment by xiaxyi** [comment]
@katbyte Updated the code, feel free to let me know if there is anything needed. Thanks~

**Comment by xiaxyi** [comment]
update acc test:
--- PASS: TestAccServiceBusSubscription_clientAffineEnabled (852.69s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/servicebus    853.419s

File: internal/services/servicebus/servicebus_subscription_resource.go [hunk 6]
```diff
@@ -129,6 +130,39 @@ func resourceServicebusSubscriptionSchema() map[string]*pluginsdk.Schema {
...
+			Type:     pluginsdk.TypeList,
+			Optional: true,
+			MaxItems: 1,
+			Elem: &pluginsdk.Resource{
+				Schema: map[string]*pluginsdk.Schema{
+					"client_id": {
+						Type:     pluginsdk.TypeString,
+						Optional: true,
+						ForceNew: true,
+					},
```
Comment (katbyte) [inline internal/services/servicebus/servicebus_subscription_resource.go:150]: should enabled be in the block?
```suggestion

		"client_scoped_subscription": {
			Type:     pluginsdk.TypeList,
			Optional: true,
			MaxItems: 1,
			Elem: &pluginsdk.Resource{
				Schema: map[string]*pluginsdk.Schema{
				
						"enabled": {
			Type:     pluginsdk.TypeBool,
			Optional: true,
			Default:  false,
		},
		
					"client_id": {
						Type:     pluginsdk.TypeString,
						Optional: true,
						ForceNew: true,
					},
```

**Review by katbyte — CHANGES_REQUESTED** [review]

File: internal/services/servicebus/servicebus_subscription_resource.go (hunk 6 above)
Comment (xiaxyi) [reply internal/services/servicebus/servicebus_subscription_resource.go:150]: @katbyte Thanks for the comment, I left it outside of the block as I was considering that the two properties are better to exist parallel? Users only need to fill the client scoped properties if they choose to enable the feature.

File: internal/services/servicebus/servicebus_subscription_resource.go [hunk 7]
```diff
@@ -140,9 +174,27 @@ func resourceServiceBusSubscriptionCreateUpdate(d *pluginsdk.ResourceData, meta
 	log.Printf("[INFO] preparing arguments for ServiceBus Subscription creation.")
 
 	var id subscriptions.Subscriptions2Id
+	name := d.Get("name").(string)
+
+	clientId := ""
+	var isSubShared bool
+	isClintAffineEnabled := d.Get("client_scoped_subscription_enabled").(bool)
+	if isClintAffineEnabled {
```
Comment (katbyte) [inline internal/services/servicebus/servicebus_subscription_resource.go:182]: could we use variables that match schema?

**Review by katbyte — CHANGES_REQUESTED** [review]

**Comment by xiaxyi** [comment]
Thanks @katbyte updated the code

**Review by katbyte — APPROVED** [review]
LGTM 🍔

**Pull request merged**